### PR TITLE
Regex collection cleanup

### PR DESCRIPTION
**Pull request by justinvp**
As part of implementing #271 (Regex collections should implement generic collection interfaces), I cleaned up some of the internals of the Regex collections. Since #271 adds new APIs, it must go in the 'future' branch. Instead of doing all the changes in 'future', I split up the changes into two parts:

 1. (This PR) **Internal cleanup** (no new APIs) to be checked-in to 'master'
 2. **Implement generic collection interfaces** (new APIs) to be checked-in to 'future'

This PR is part 1. Once this PR has been reviewed, accepted, and merged into 'master', please merge 'master' into 'future'. Then I'll submit another PR in 'future' that adds the generic interfaces, building on top of these changes.

### Summary of changes in this PR:
 - `MatchCollection` enumerator tests
 - Fields: internal -> private
 - Made enumerators private nested classes
 - Reduced the size of `MatchCollection`'s enumerator
 - Enumerator field naming improvements
 - Reorganized member order (prep for generic collection members)
 - Removed unnecessary static field (`s_infinite`) from `MatchCollection`
 - Use language keywords instead of BCL types

### CODE REVIEW

File: src/System.Text.RegularExpressions/tests/MatchCollectionTests.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            var regex = new Regex("e");
+            var collection = regex.Matches("dotnet");
+            var enumerator = collection.GetEnumerator();
+
+            Assert.Throws<InvalidOperationException>(() => enumerator.Current);
+            Assert.True(enumerator.MoveNext());
+            Assert.IsAssignableFrom<Match>(enumerator.Current);
+            Assert.Equal(4, ((Match)enumerator.Current).Index);
+            Assert.Equal("e", ((Match)enumerator.Current).Value);
+            Assert.False(enumerator.MoveNext());
```
Comment (Maxwe11) [inline src/System.Text.RegularExpressions/tests/MatchCollectionTests.cs]: `Assert.Empty`

File: src/System.Text.RegularExpressions/tests/MatchCollectionTests.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            Assert.Throws<InvalidOperationException>(() => enumerator.Current);
+
+            enumerator.Reset();
+            Assert.Throws<InvalidOperationException>(() => enumerator.Current);
+            Assert.True(enumerator.MoveNext());
```
Comment (Maxwe11) [inline src/System.Text.RegularExpressions/tests/MatchCollectionTests.cs]: `Asset.NotEmpty`

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by ellismg** [comment]
LGTM

**Comment by stephentoub** [comment]
LGTM

**Pull request merged**